File: src/main/java/com/clinica/atendimento/dto/TerapiaDTO.java
package com.clinica.atendimento.dto;

import com.clinica.atendimento.model.Terapia;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Accessors(fluent = true)
@Builder
public class TerapiaDTO {

    @JsonProperty
    private Long id;
    @JsonProperty
    private String nome;

    // Static method to convert from entity to DTO
    public static TerapiaDTO fromEntity(Terapia terapia) {
        if (terapia == null) {
            return null;
        }

        return TerapiaDTO.builder()
                .id(terapia.id())
                .nome(terapia.nome())
                .build();
    }

    // Method to convert from DTO to entity
    public Terapia toEntity() {
        // Using the constructor for backward compatibility
        // return new Terapia(id, nome);

        // Using the builder pattern
        Terapia.TerapiaBuilder builder = Terapia.builder();

        // Only set ID if it's not null
        if (id != null) {
            builder.id(id);
        }

        return builder
                .nome(nome)
                .build();
    }
}

File: src/main/kotlin/com/clinica/atendimento/dto/TerapiaDTO.kt
package com.clinica.atendimento.dto

import com.clinica.atendimento.model.Terapia
import com.fasterxml.jackson.annotation.JsonProperty

data class TerapiaDTO(
    @JsonProperty
    var id: Long? = null,
    
    @JsonProperty
    var nome: String? = null
) {
    // Fluent accessors to maintain compatibility with Java code
    fun id(id: Long?): TerapiaDTO {
        this.id = id
        return this
    }

    fun nome(nome: String?): TerapiaDTO {
        this.nome = nome
        return this
    }

    // Method to convert from DTO to entity
    fun toEntity(): Terapia {
        val builder = Terapia.builder()

        // Only set ID if it's not null
        id?.let { builder.id(it) }

        return builder
            .nome(nome)
            .build()
    }

    companion object {
        // Static method to convert from entity to DTO
        @JvmStatic
        fun fromEntity(terapia: Terapia?): TerapiaDTO? {
            if (terapia == null) {
                return null
            }

            return TerapiaDTO(
                id = terapia.id,
                nome = terapia.nome
            )
        }

        @JvmStatic
        fun builder(): TerapiaDTOBuilder {
            return TerapiaDTOBuilder()
        }
    }

    class TerapiaDTOBuilder {
        private val instance = TerapiaDTO()

        fun id(id: Long?): TerapiaDTOBuilder {
            instance.id = id
            return this
        }

        fun nome(nome: String?): TerapiaDTOBuilder {
            instance.nome = nome
            return this
        }

        fun build(): TerapiaDTO {
            return instance
        }
    }
}